upstream/man/tcumttest.1
.TH "TCUMTTEST" 1 "2010-08-05" "Man Page" "Tokyo Cabinet"

.SH NAME
tcutest \- test cases of the on-memory database API

.SH DESCRIPTION
.PP
The command `\fBtcumttest\fR' is a utility for facility test under multi\-thread situation.  This command is used in the following format.  `\fIrnum\fR' specifies the number of iterations.  `\fIbnum\fR' specifies the number of buckets.
.PP
.RS
.br
\fBtcumttest combo \fR[\fB\-rnd\fR]\fB \fItnum\fB \fIrnum\fB \fR[\fB\fIbnum\fB\fR]\fB\fR
.RS
Peform storing, retrieving, and removing in turn.
.RE
.br
\fBtcumttest typical \fR[\fB\-nc\fR]\fB \fR[\fB\-rr \fInum\fB\fR]\fB \fItnum\fB \fIrnum\fB \fR[\fB\fIbnum\fB\fR]\fB\fR
.RS
Perform typical operations selected at random.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-rnd\fR : select keys at random.
.br
\fB\-nc\fR : omit the comparison test.
.br
\-rr \fInum\fR : specifiy the ratio of reading operation by percentage.
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR tcutest (1),
.BR tcucodec (1),
.BR tcutil (3),
.BR tokyocabinet (3)


upstream/ChangeLog
2009-12-07  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbidxputtoken, tctdbidxputqgram): domain overflow checking was added.

	- Release: 1.4.41

2009-11-24  Mikio Hirabayashi  <hirarin@gmail.com>

	* tchdb.c (tchdbputimpl): efficiency of using the free block pool was improved.

	* tchdb.c (tchdboptimizeimpl): the behaviour in no lock mode was modified.

	- Release: 1.4.40

2009-11-11  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.h (TCALIGNOF, tcgeneric_t): new macros.

	* tcutil.h (TCALIGNPAD): the alignment is now calculated by the generic union.

	* myconf.h (_alignof, _issigned, _maxof): new macros.

	- Release: 1.4.39

2009-11-09  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcsysinfo): FreeBSD and Mac OS X are now supported.

	* tcutil.c (tcmpoolclear): new function.

	* tchdb.c (tchdbputproc): a potential bug of memory aligment violation was fixed.

	* tcbdb.c (tcbdbputproc): a potential bug of memory aligment violation was fixed.

	* tcfdb.c (tcfdbputproc): a potential bug of memory aligment violation was fixed.

	- Release: 1.4.38

2009-10-27  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbcacheclear, tctdbcacheclearimpl): new functions.

	* tcadb.c (tcadbmisc): "getpart", "cacheclear", and "regex" functions were added.

	- Release: 1.4.37

2009-10-26  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcwwwformdecode2): a bug related to multiple cookie handling was fixed.

	* tcadb.c (tcadbmisc): the "error" functions were added.

	- Release: 1.4.36

2009-10-06  Mikio Hirabayashi  <hirarin@gmail.com>

	* tchdb.c (tchdbseekwrite): a bug related to emulation of the UBC was fixed.

	* tchdb.c (tchdbvanishimpl): potential memory leak was removed.

	* tcfdb.c (tcfdbvanishimpl): potential memory leak was removed.

	* tcadb.c (tcadbsetskelmulti, tcadbmulnew, tcadbmuldel): new functions.

	* tcadb.c (tcadbmisc): "range" sub function for B+ tree was added.

	* tcamgr.c (setskeltran): new function.

	* tcamttest.c: new file.

	- Release: 1.4.35

2009-09-11  Mikio Hirabayashi  <hirarin@gmail.com>

	* myconf.h: a macro for environment without "nanl" was added.

	* tctdbmetasearch (tctdbmetasearch): a bug related to the limit parameters was fixed.

	- Release: 1.4.34

2009-08-31  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.h, tchdb.h, tcbdb.h, tcfdb.h, tctdb.h, tcadb.h: including stdbool.h was wrapped.

	* tcutil.c (tccmpdecimal): accuracy was improved and now real number is supported.

	* tctdb.c (tctdbqrycondmatch): accuracy of decimal operators was improved.

	- Release: 1.4.33

2009-08-14  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tctmpldumpeval): "INC", "PRT" operators and "SET" directive were added.

	* tcutil.c (tcwwwformdecode2): a bug of memory corruption was fixed.

	* tchdb.c (tchdbgetimpl): a bug related to caching large records was fixed.

	- Release: 1.4.32

2009-07-23  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcmpoolpop, tcstatfile): new functions.

	* tcutil.c (tcwwwformdecode2, tcarccipher): new functions.

	* tcutil.c (tctmpldumpeval, tctmpldumpevalvar): type checking was improved.

	- Release: 1.4.31

2009-07-13  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbstrtometasearcytype): new function.

	* tcadb.c (tcadbmisc): "metasearch" command was added.

	- Release: 1.4.30

2009-07-03  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbidxgetbyftsunion): efficiency counting sort was improved.

	- Release: 1.4.29

2009-06-22  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcstrutfnorm, tcstrkwic, tcstrtokenize): new functions.

	* tcutil.c (tcstrutfnorm, tcstrucsnorm): "TCUNWIDTH" option was added.

	* tctdb.c (tctdbsetinvcache, tctdbidxsyncicc, tctdbidxcmpkey): new functions.

	* tctdb.c (tctdbmetasearch, tctdbget4, tctdbqrykwic): new functions.

	* tctdb.c (tctdbsetindeximpl): the xmsiz parameter is now configured.

	- Release: 1.4.28

2009-06-16  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcstrskipspc, tcstrucsnorm): new functions.

	* tctdb.c (tctdbqryaddcond): full-text search operators were added.

	* tctdb.c (tctdbsetindex): q-gram inverted index was added.

	* tctdb.c (tctdbidxputqgram, tctdbidxoutqgram, tctdbidxgetbyfts): new functions.

	- Release: 1.4.27

2009-06-13  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbsetindex, tctdbsearchimpl): token inverted index was added.

	* tctdb.c (tctdbidxputone, tctdbidxoutone): new functions.

	* tctdb.c (tctdbidxputtoken, tctdbidxouttoken, tctdbidxgetbytokens): new functions.

	* tcadb.c (tcadbput, tcadbputkeep, tcadbputcat): capnum behaviour was modified.

	- Release: 1.4.26

2009-06-09  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tccstrescape, tccstrunescape, tcjsonescape, tcjsonunescape): new functions.

	* tcutil.c (tcpathlock, tcpathunlock): new functions.

	* tchdb.c (tchdbopen): double opening detection was added.

	* tchdb.c (tchdbtranbegin): a bug of invalid flag setting was fixed.

	* tcfdb.c (tcfdbopen): double opening detection was added.

	* tcfdb.c (tcfdbtranbegin): a bug of invalid flag setting was fixed.

	* tctdb.c (tctdbiternext3): new function.

	- Release: 1.4.25

2009-06-04  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcwwwformencode, tcwwwformdecode): new functions.

	* tcutil.c (tctmplnew, tctmpldel, tctmplload, tctmpldump): new functions.

	* tcutil.c (tcmapget4, tctreeget4): new functions.

	* tcutil.c (tclistprintf, tcmapprintf, tctreeprintf): new functions.

	* tcucodec.c (runtmpl, proctmpl): new functions.

	- Release: 1.4.24

2009-05-24  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcmapiterinit2, tcmdbiterinit2): new functions.

	* tchdb.c (tchdbiterinit2, tchdbiterjumpimpl): new functions.

	* tcfdb.c (tcfdbiterinit2, tcfdbiterjumpimpl): new functions.

	* tctdb.c (tctdbiterinit2): new function.

	- Release: 1.4.23

2009-05-21  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcadb.c (tcadbmisc): "sync", "optimize", "vanish" commands were added.

	- Release: 1.4.22


upstream/tcumttest.c
/*************************************************************************************************
 * The test cases of the on-memory database API
 *                                                      Copyright (C) 2006-2008 Mikio Hirabayashi
 * This file is part of Tokyo Cabinet.
 * Tokyo Cabinet is free software; you can redistribute it and/or modify it under the terms of
 * the GNU Lesser General Public License as published by the Free Software Foundation; either
 * version 2.1 of the License or any later version.  Tokyo Cabinet is distributed in the hope
 * that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 * You should have received a copy of the GNU Lesser General Public License along with Tokyo
 * Cabinet; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA.
 *************************************************************************************************/


#include <tcutil.h>
#include "myconf.h"

#define RECBUFSIZ      32                // buffer for records

typedef struct {                         // type of structure for combo thread
  TCMDB *mdb;
  int rnum;
  bool rnd;
  int id;
} TARGCOMBO;

typedef struct {                         // type of structure for typical thread
  TCMDB *mdb;
  int rnum;
  bool nc;
  int rratio;
  int id;
} TARGTYPICAL;


/* global variables */
const char *g_progname;                  // program name


/* function prototypes */
int main(int argc, char **argv);
static void usage(void);
static void iprintf(const char *format, ...);
static void eprint(const char *func);
static int myrand(int range);
static int myrandnd(int range);
static int runcombo(int argc, char **argv);
static int runtypical(int argc, char **argv);
static int proccombo(int tnum, int rnum, int bnum, bool rnd);
static int proctypical(int tnum, int rnum, int bnum, bool nc, int rratio);
static void *threadwrite(void *targ);
static void *threadread(void *targ);
static void *threadremove(void *targ);
static void *threadtypical(void *targ);


/* main routine */
int main(int argc, char **argv){
  g_progname = argv[0];
  srand((unsigned int)(tctime() * 1000) % UINT_MAX);
  if(argc < 2) usage();
  int rv = 0;
  if(!strcmp(argv[1], "combo")){
    rv = runcombo(argc, argv);
  } else if(!strcmp(argv[1], "typical")){
    rv = runtypical(argc, argv);
  } else {
    usage();
  }
  return rv;
}


/* print the usage and exit */
static void usage(void){
  fprintf(stderr, "%s: test cases of the on-memory database API of Tokyo Cabinet\n", g_progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "usage:\n");
  fprintf(stderr, "  %s combo [-rnd] tnum rnum [bnum]\n", g_progname);
  fprintf(stderr, "  %s typical [-nc] [-rr num] tnum rnum [bnum]\n", g_progname);
  fprintf(stderr, "\n");
  exit(1);
}


/* print formatted information string and flush the buffer */
static void iprintf(const char *format, ...){
  va_list ap;
  va_start(ap, format);
  vprintf(format, ap);
  fflush(stdout);
  va_end(ap);
}


/* print error message of on-memory database */
static void eprint(const char *func){
  fprintf(stderr, "%s: %s: error\n", g_progname, func);
}


/* get a random number */
static int myrand(int range){
  return (int)((double)range * rand() / (RAND_MAX + 1.0));
}


/* get a random number based on normal distribution */
static int myrandnd(int range){
  int num = (int)tcdrandnd(range >> 1, range / 10);
  return (num < 0 || num >= range) ? 0 : num;
}


/* parse arguments of combo command */
static int runcombo(int argc, char **argv){
  char *tstr = NULL;
  char *rstr = NULL;
  char *bstr = NULL;
  bool rnd = false;
  for(int i = 2; i < argc; i++){
    if(!tstr && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-rnd")){
        rnd = true;
      } else {
        usage();
      }
    } else if(!tstr){
      tstr = argv[i];
    } else if(!rstr){
      rstr = argv[i];
    } else if(!bstr){
      bstr = argv[i];
    } else {
      usage();
    }
  }
  if(!tstr || !rstr) usage();
  int tnum = atoi(tstr);
  int rnum = atoi(rstr);
  if(tnum < 1 || rnum < 1) usage();
  int bnum = bstr ? atoi(bstr) : -1;
  int rv = proccombo(tnum, rnum, bnum, rnd);
  return rv;
}


/* parse arguments of typical command */
static int runtypical(int argc, char **argv){
  char *tstr = NULL;
  char *rstr = NULL;
  char *bstr = NULL;
  int rratio = -1;
  bool nc = false;
  for(int i = 2; i < argc; i++){
    if(!tstr && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-nc")){
        nc = true;
      } else if(!strcmp(argv[i], "-rr")){
        if(++i >= argc) usage();
        rratio = atoi(argv[i]);
      } else {
        usage();
      }
    } else if(!tstr){
      tstr = argv[i];
    } else if(!rstr){
      rstr = argv[i];
    } else if(!bstr){
      bstr = argv[i];
    } else {
      usage();
    }
  }
  if(!tstr || !rstr) usage();
  int tnum = atoi(tstr);
  int rnum = atoi(rstr);
  if(tnum < 1 || rnum < 1) usage();
  int bnum = bstr ? atoi(bstr) : -1;
  int rv = proctypical(tnum, rnum, bnum, nc, rratio);
  return rv;
}


/* perform combo command */
static int proccombo(int tnum, int rnum, int bnum, bool rnd){
  iprintf("<Combination Test>\n  tnum=%d  rnum=%d  bnum=%d  rnd=%d\n\n", tnum, rnum, bnum, rnd);
  bool err = false;
  double stime = tctime();
  TCMDB *mdb = (bnum > 0) ? tcmdbnew2(bnum) : tcmdbnew();
  TARGCOMBO targs[tnum];
  pthread_t threads[tnum];
  if(tnum == 1){
    targs[0].mdb = mdb;
    targs[0].rnum = rnum;
    targs[0].rnd = rnd;
    targs[0].id = 0;
    if(threadwrite(targs) != NULL) err = true;
  } else {
    for(int i = 0; i < tnum; i++){
      targs[i].mdb = mdb;
      targs[i].rnum = rnum;
      targs[i].rnd = rnd;
      targs[i].id = i;
      if(pthread_create(threads + i, NULL, threadwrite, targs + i) != 0){
        eprint("pthread_create");
        targs[i].id = -1;
        err = true;
      }
    }
    for(int i = 0; i < tnum; i++){
      if(targs[i].id == -1) continue;
      void *rv;
      if(pthread_join(threads[i], &rv) != 0){
        eprint("pthread_join");
        err = true;
      } else if(rv){
        err = true;
      }
    }
  }
  if(tnum == 1){
    targs[0].mdb = mdb;
    targs[0].rnum = rnum;
    targs[0].rnd = rnd;
    targs[0].id = 0;
    if(threadread(targs) != NULL) err = true;
  } else {
    for(int i = 0; i < tnum; i++){
      targs[i].mdb = mdb;
      targs[i].rnum = rnum;
      targs[i].rnd = rnd;
      targs[i].id = i;
      if(pthread_create(threads + i, NULL, threadread, targs + i) != 0){
        eprint("pthread_create");
        targs[i].id = -1;
        err = true;
      }
    }
    for(int i = 0; i < tnum; i++){
      if(targs[i].id == -1) continue;
      void *rv;
      if(pthread_join(threads[i], &rv) != 0){
        eprint("pthread_join");
        err = true;
      } else if(rv){
        err = true;
      }
    }
  }
  if(tnum == 1){
    targs[0].mdb = mdb;
    targs[0].rnum = rnum;
    targs[0].rnd = rnd;
    targs[0].id = 0;
    if(threadremove(targs) != NULL) err = true;
  } else {
    for(int i = 0; i < tnum; i++){
      targs[i].mdb = mdb;
      targs[i].rnum = rnum;
      targs[i].rnd = rnd;
      targs[i].id = i;
      if(pthread_create(threads + i, NULL, threadremove, targs + i) != 0){
        eprint("pthread_create");
        targs[i].id = -1;
        err = true;
      }
    }
    for(int i = 0; i < tnum; i++){
      if(targs[i].id == -1) continue;
      void *rv;
      if(pthread_join(threads[i], &rv) != 0){
        eprint("pthread_join");
        err = true;
      } else if(rv){
        err = true;
      }
    }
  }
  iprintf("record number: %llu\n", (unsigned long long)tcmdbrnum(mdb));
  tcmdbdel(mdb);
  iprintf("time: %.3f\n", tctime() - stime);
  iprintf("%s\n\n", err ? "error" : "ok");
  return err ? 1 : 0;
}


/* perform typical command */
static int proctypical(int tnum, int rnum, int bnum, bool nc, int rratio){
  iprintf("<Typical Access Test>\n  tnum=%d  rnum=%d  bnum=%d  nc=%d  rratio=%d\n\n",
          tnum, rnum, bnum, nc, rratio);
  bool err = false;
  double stime = tctime();
  TCMDB *mdb = (bnum > 0) ? tcmdbnew2(bnum) : tcmdbnew();
  TARGTYPICAL targs[tnum];
  pthread_t threads[tnum];
  if(tnum == 1){
    targs[0].mdb = mdb;
    targs[0].rnum = rnum;
    targs[0].nc = nc;
    targs[0].rratio = rratio;
    targs[0].id = 0;
    if(threadtypical(targs) != NULL) err = true;
  } else {
    for(int i = 0; i < tnum; i++){
      targs[i].mdb = mdb;
      targs[i].rnum = rnum;
      targs[i].nc = nc;
      targs[i].rratio= rratio;
      targs[i].id = i;
      if(pthread_create(threads + i, NULL, threadtypical, targs + i) != 0){
        eprint("pthread_create");
        targs[i].id = -1;
        err = true;
      }
    }
    for(int i = 0; i < tnum; i++){
      if(targs[i].id == -1) continue;
      void *rv;
      if(pthread_join(threads[i], &rv) != 0){
        eprint("pthread_join");
        err = true;
      } else if(rv){
        err = true;
      }
    }
  }
  iprintf("record number: %llu\n", (unsigned long long)tcmdbrnum(mdb));
  tcmdbdel(mdb);
  iprintf("time: %.3f\n", tctime() - stime);
  iprintf("%s\n\n", err ? "error" : "ok");
  return err ? 1 : 0;
}


/* thread the write function */
static void *threadwrite(void *targ){
  TCMDB *mdb = ((TARGCOMBO *)targ)->mdb;
  int rnum = ((TARGCOMBO *)targ)->rnum;
  bool rnd = ((TARGCOMBO *)targ)->rnd;
  int id = ((TARGCOMBO *)targ)->id;
  double stime = tctime();
  if(id == 0) iprintf("writing:\n");
  int base = id * rnum;
  for(int i = 1; i <= rnum; i++){
    char buf[RECBUFSIZ];
    int len = sprintf(buf, "%08d", base + (rnd ? myrand(i) : i));
    tcmdbput(mdb, buf, len, buf, len);
    if(id == 0 && rnum > 250 && i % (rnum / 250) == 0){
      putchar('.');
      fflush(stdout);
      if(i == rnum || i % (rnum / 10) == 0) iprintf(" (%08d)\n", i);
    }
  }
  if(id == 0) iprintf("time: %.3f\n", tctime() - stime);
  return NULL;
}


/* thread the read function */
static void *threadread(void *targ){
  TCMDB *mdb = ((TARGCOMBO *)targ)->mdb;
  int rnum = ((TARGCOMBO *)targ)->rnum;
  bool rnd = ((TARGCOMBO *)targ)->rnd;
  int id = ((TARGCOMBO *)targ)->id;
  double stime = tctime();
  if(id == 0) iprintf("reading:\n");
  int base = id * rnum;
  for(int i = 1; i <= rnum; i++){
    char kbuf[RECBUFSIZ];
    int ksiz = sprintf(kbuf, "%08d", base + (rnd ? myrand(i) : i));
    int vsiz;
    char *vbuf = tcmdbget(mdb, kbuf, ksiz, &vsiz);
    if(vbuf) tcfree(vbuf);
    if(id == 0 && rnum > 250 && i % (rnum / 250) == 0){
      putchar('.');
      fflush(stdout);
      if(i == rnum || i % (rnum / 10) == 0) iprintf(" (%08d)\n", i);
    }
  }
  if(id == 0) iprintf("time: %.3f\n", tctime() - stime);
  return NULL;
}


/* thread the remove function */
static void *threadremove(void *targ){
  TCMDB *mdb = ((TARGCOMBO *)targ)->mdb;
  int rnum = ((TARGCOMBO *)targ)->rnum;
  bool rnd = ((TARGCOMBO *)targ)->rnd;
  int id = ((TARGCOMBO *)targ)->id;
  double stime = tctime();
  if(id == 0) iprintf("removing:\n");
  int base = id * rnum;
  for(int i = 1; i <= rnum; i++){
    char buf[RECBUFSIZ];
    int len = sprintf(buf, "%08d", base + (rnd ? myrand(i) : i));
    tcmdbout(mdb, buf, len);
    if(id == 0 && rnum > 250 && i % (rnum / 250) == 0){
      putchar('.');
      fflush(stdout);
      if(i == rnum || i % (rnum / 10) == 0) iprintf(" (%08d)\n", i);
    }
  }
  if(id == 0) iprintf("time: %.3f\n", tctime() - stime);
  return NULL;
}


/* thread the typical function */
static void *threadtypical(void *targ){
  TCMDB *mdb = ((TARGTYPICAL *)targ)->mdb;
  int rnum = ((TARGTYPICAL *)targ)->rnum;
  bool nc = ((TARGTYPICAL *)targ)->nc;
  int rratio = ((TARGTYPICAL *)targ)->rratio;
  int id = ((TARGTYPICAL *)targ)->id;
  bool err = false;
  TCMAP *map = (!nc && id == 0) ? tcmapnew2(rnum + 1) : NULL;
  int base = id * rnum;
  int mrange = tclmax(50 + rratio, 100);
  for(int i = 1; !err && i <= rnum; i++){
    char buf[RECBUFSIZ];
    int len = sprintf(buf, "%08d", base + myrandnd(i));
    int rnd = myrand(mrange);
    if(rnd < 10){
      tcmdbput(mdb, buf, len, buf, len);
      if(map) tcmapput(map, buf, len, buf, len);
    } else if(rnd < 15){
      tcmdbputkeep(mdb, buf, len, buf, len);
      if(map) tcmapputkeep(map, buf, len, buf, len);
    } else if(rnd < 20){
      tcmdbputcat(mdb, buf, len, buf, len);
      if(map) tcmapputcat(map, buf, len, buf, len);
    } else if(rnd < 30){
      tcmdbout(mdb, buf, len);
      if(map) tcmapout(map, buf, len);
    } else if(rnd < 31){
      if(myrand(10) == 0) tcmdbiterinit(mdb);
      for(int j = 0; !err && j < 10; j++){
        int ksiz;
        char *kbuf = tcmdbiternext(mdb, &ksiz);
        if(kbuf) tcfree(kbuf);
      }
    } else {
      int vsiz;
      char *vbuf = tcmdbget(mdb, buf, len, &vsiz);
      if(vbuf){
        if(map){
          int msiz;
          const char *mbuf = tcmapget(map, buf, len, &msiz);
          if(msiz != vsiz || memcmp(mbuf, vbuf, vsiz)){
            eprint("(validation)");
            err = true;
          }
        }
        tcfree(vbuf);
      } else {
        if(map && tcmapget(map, buf, len, &vsiz)){
          eprint("(validation)");
          err = true;
        }
      }
    }
    if(id == 0 && rnum > 250 && i % (rnum / 250) == 0){
      putchar('.');
      fflush(stdout);
      if(i == rnum || i % (rnum / 10) == 0) iprintf(" (%08d)\n", i);
    }
  }
  if(map){
    tcmapiterinit(map);
    int ksiz;
    const char *kbuf;
    while(!err && (kbuf = tcmapiternext(map, &ksiz)) != NULL){
      int vsiz;
      char *vbuf = tcmdbget(mdb, kbuf, ksiz, &vsiz);
      if(vbuf){
        int msiz;
        const char *mbuf = tcmapget(map, kbuf, ksiz, &msiz);
        if(!mbuf || msiz != vsiz || memcmp(mbuf, vbuf, vsiz)){
          eprint("(validation)");
          err = true;
        }
        tcfree(vbuf);
      } else {
        eprint("(validation)");
        err = true;
      }
    }
    tcmapdel(map);
  }
  return err ? "error" : NULL;
}



// END OF FILE
